File: manager.cpp
#include "manager.hpp"
#include <iostream>
#include <iomanip>

Manager::Manager()
	: Employee()
{
	// Устанавливаем должность менеджера
	SetPosition("Менеджер");
}

Manager::Manager(const std::string& fullName, int age, const std::string& contactNumber,
				 const std::string& address, double hourlyRate,
				 const std::string& login, const std::string& password)
	: Employee(fullName, age, contactNumber, address, "Менеджер", hourlyRate, login, password)
{
}

void Manager::AddPenalty(Employee* employee, double amount)
{
	if (employee != nullptr && amount > 0.0)
	{
		double currentBalance = employee->GetSalaryBalance();
		employee->SetSalaryBalance(currentBalance - amount);	// Штраф вычитается из баланса
		std::cout << "Штраф в размере " << amount 
				  << " руб. добавлен сотруднику: " << employee->GetFullName() << std::endl;
	}
}

void Manager::AddBonus(Employee* employee, double amount)
{
	if (employee != nullptr && amount > 0.0)
	{
		double currentBalance = employee->GetSalaryBalance();
		employee->SetSalaryBalance(currentBalance + amount);	// Премия добавляется к балансу
		std::cout << "Премия в размере " << amount 
				  << " руб. добавлена сотруднику: " << employee->GetFullName() << std::endl;
	}
}

void Manager::ChangeHourlyRate(Employee* employee, double newRate)
{
	if (employee != nullptr && newRate >= 0.0)
	{
		employee->SetHourlyRate(newRate);
		std::cout << "Почасовая ставка изменена для сотрудника: " 
				  << employee->GetFullName() << " на " << newRate << " руб./час" << std::endl;
	}
}

void Manager::ChangeHoursWorked(Employee* employee, double hours)
{
	if (employee != nullptr && hours >= 0.0)
	{
		employee->SetHoursWorked(hours);
		std::cout << "Отработанные часы изменены для сотрудника: " 
				  << employee->GetFullName() << " на " << hours << " часов" << std::endl;
	}
}

void Manager::ResetHoursWorked(Employee* employee)
{
	if (employee != nullptr)
	{
		employee->SetHoursWorked(0.0);
		std::cout << "Отработанные часы сброшены для сотрудника: " 
				  << employee->GetFullName() << std::endl;
	}
}

void Manager::ChangeSellingPrice(MenuItem* menuItem, double newPrice)
{
	if (menuItem != nullptr && newPrice >= 0.0)
	{
		menuItem->SetSellingPrice(newPrice);
		std::cout << "Цена продажи изменена для позиции '" << menuItem->GetName() 
				  << "' на " << newPrice << " руб." << std::endl;
	}
}

MenuItem* Manager::CreateMenuItem(const std::string& name, double costPrice,
								  double sellingPrice, bool isAvailable)
{
	MenuItem* newItem = new MenuItem(name, costPrice, sellingPrice, isAvailable);
	std::cout << "Создана новая позиция в меню: " << name << std::endl;
	return newItem;
}

void Manager::EditMenuItem(MenuItem* menuItem, const std::string& newName,
						   double newCostPrice, double newSellingPrice)
{
	if (menuItem != nullptr)
	{
		menuItem->SetName(newName);
		menuItem->SetCostPrice(newCostPrice);
		menuItem->SetSellingPrice(newSellingPrice);
		std::cout << "Позиция меню отредактирована: " << newName << std::endl;
	}
}

void Manager::ViewSalesReport(const MenuItemsType& menuItems) const
{
	std::cout << "=== Отчет по продажам ===" << std::endl;
	std::cout << "Отчет создан менеджером: " << GetFullName() << std::endl;
	std::cout << std::endl;

	if (menuItems.empty())
	{
		std::cout << "Меню пусто." << std::endl;
	}
	else
	{
		double totalProfit = 0.0;
		int totalSales = 0;

		for (const auto* item : menuItems)
		{
			if (item != nullptr)
			{
				int sales = item->GetSalesCount();
				double profit = item->CalculateTotalProfit();
				totalSales += sales;
				totalProfit += profit;

				std::cout << "Позиция: " << item->GetName() 
						  << " | Продаж: " << sales 
						  << " | Прибыль: " << std::fixed << std::setprecision(2) 
						  << profit << " руб." << std::endl;
			}
		}

		std::cout << std::endl;
		std::cout << "Итого продаж: " << totalSales << std::endl;
		std::cout << "Общая прибыль: " << totalProfit << " руб." << std::endl;
	}

	std::cout << "=========================" << std::endl;
}

void Manager::ManageInventory(Inventory* inventory)
{
	if (inventory != nullptr)
	{
		std::cout << "Управление складом менеджером: " << GetFullName() << std::endl;
		inventory->PrintInventoryReport();
		std::cout << std::endl;
		inventory->PrintExpiryDates();
	}
}

File: lab4/Manager.java
import java.util.ArrayList;

// Класс менеджера, наследуется от Employee и реализует множественное наследование
// Наследуется от Employee и реализует интерфейсы IReportGenerator и ICloneable
// Также использует AbstractReportGenerator через композицию для демонстрации абстрактного класса
public class Manager extends Employee implements IReportGenerator, ICloneable<Manager> {
    // Тип для списка сотрудников
    public static class EmployeesType extends ArrayList<Employee> {
        // Наследуемся от ArrayList для удобства работы
    }

    // Тип для списка позиций меню
    public static class MenuItemsType extends ArrayList<MenuItem> {
        // Наследуемся от ArrayList для удобства работы
    }

    // Тип для списка заказов
    public static class OrdersType extends ArrayList<Order> {
        // Наследуемся от ArrayList для удобства работы
    }

    // Композиция: используем AbstractReportGenerator для демонстрации абстрактного класса
    private AbstractReportGenerator reportGenerator;
    
    // Конструктор по умолчанию
    public Manager() {
        super();  // Вызов конструктора базового класса Employee
        // Устанавливаем должность менеджера
        setPosition("Менеджер");
        // Создаем объект абстрактного класса через анонимный класс
        this.reportGenerator = new AbstractReportGenerator("Менеджер") {
            @Override
            public String generateReport() {
                return Manager.this.generateReport();
            }
            
            @Override
            public String getReportType() {
                return Manager.this.getReportType();
            }
        };
    }

    // Конструктор с параметрами - демонстрация вызова конструктора базового класса
    public Manager(String fullName, int age, String contactNumber,
                  String address, double hourlyRate,
                  String login, String password) {
        // Вызов конструктора базового класса Employee с параметрами
        super(fullName, age, contactNumber, address, "Менеджер", hourlyRate, login, password);
        // Создаем объект абстрактного класса через анонимный класс
        this.reportGenerator = new AbstractReportGenerator("Менеджер") {
            @Override
            public String generateReport() {
                return Manager.this.generateReport();
            }
            
            @Override
            public String getReportType() {
                return Manager.this.getReportType();
            }
        };
    }

    // Добавить штраф сотруднику
    public void addPenalty(Employee employee, double amount) {
        if (employee != null && amount > 0.0) {
            double currentBalance = employee.getSalaryBalance();
            employee.setSalaryBalance(currentBalance - amount); // Штраф вычитается из баланса
            System.out.println("Штраф в размере " + amount + " руб. добавлен сотруднику: " + 
                             employee.getFullName());
        }
    }

    // Добавить премию сотруднику
    public void addBonus(Employee employee, double amount) {
        if (employee != null && amount > 0.0) {
            double currentBalance = employee.getSalaryBalance();
            employee.setSalaryBalance(currentBalance + amount); // Премия добавляется к балансу
            System.out.println("Премия в размере " + amount + " руб. добавлена сотруднику: " + 
                             employee.getFullName());
        }
    }

    // Изменить почасовую ставку
    public void changeHourlyRate(Employee employee, double newRate) {
        if (employee == null) {
            throw new IllegalArgumentException("Сотрудник не может быть null");
        }
        if (newRate < 0.0) {
            // Демонстрация генерации исключения
            throw new IllegalArgumentException("Нельзя установить отрицательную почасовую ставку");
        }
        employee.setHourlyRate(newRate);
        System.out.println("Почасовая ставка изменена для сотрудника: " + 
                         employee.getFullName() + " на " + newRate + " руб./час");
    }

    // Изменить отработанные часы
    public void changeHoursWorked(Employee employee, double hours) {
        if (employee != null && hours >= 0.0) {
            employee.setHoursWorked(hours);
            System.out.println("Отработанные часы изменены для сотрудника: " + 
                             employee.getFullName() + " на " + hours + " часов");
        }
    }

    // Сбросить отработанные часы
    public void resetHoursWorked(Employee employee) {
        if (employee != null) {
            employee.setHoursWorked(0.0);
            System.out.println("Отработанные часы сброшены для сотрудника: " + 
                             employee.getFullName());
        }
    }

    // Изменить цену продажи
    public void changeSellingPrice(MenuItem menuItem, double newPrice) {
        if (menuItem != null && newPrice >= 0.0) {
            menuItem.setSellingPrice(newPrice);
            System.out.println("Цена продажи изменена для позиции '" + menuItem.getName() + 
                             "' на " + newPrice + " руб.");
        }
    }

    // Создать позицию в меню
    public MenuItem createMenuItem(String name, double costPrice,
                                   double sellingPrice, boolean isAvailable) {
        MenuItem newItem = new MenuItem(name, costPrice, sellingPrice, isAvailable);
        System.out.println("Создана новая позиция в меню: " + name);
        return newItem;
    }

    // Редактировать позицию в меню
    public void editMenuItem(MenuItem menuItem, String newName,
                            double newCostPrice, double newSellingPrice) {
        if (menuItem != null) {
            menuItem.setName(newName);
            menuItem.setCostPrice(newCostPrice);
            menuItem.setSellingPrice(newSellingPrice);
            System.out.println("Позиция меню отредактирована: " + newName);
        }
    }

    // Просмотр отчетов по продажам по позициям меню
    public void viewSalesReport(MenuItemsType menuItems) {
        System.out.println("=== Отчет по продажам ===");
        System.out.println("Отчет создан менеджером: " + getFullName());
        System.out.println();

        if (menuItems.isEmpty()) {
            System.out.println("Меню пусто.");
        } else {
            double totalProfit = 0.0;
            int totalSales = 0;

            for (MenuItem item : menuItems) {
                if (item != null) {
                    int sales = item.getSalesCount();
                    double profit = item.calculateTotalProfit();
                    totalSales += sales;
                    totalProfit += profit;

                    System.out.printf("Позиция: %s | Продаж: %d | Прибыль: %.2f руб.%n", 
                                    item.getName(), sales, profit);
                }
            }

            System.out.println();
            System.out.println("Итого продаж: " + totalSales);
            System.out.printf("Общая прибыль: %.2f руб.%n", totalProfit);
        }

        System.out.println("=========================");
    }

    // Просмотр отчетов по заказам
    public void viewOrdersReport(OrdersType orders) {
        System.out.println("=== Отчет по заказам ===");
        System.out.println("Отчет создан менеджером: " + getFullName());
        System.out.println();

        if (orders.isEmpty()) {
            System.out.println("Заказов нет.");
        } else {
            double totalRevenue = 0.0;
            int totalOrders = 0;

            for (Order order : orders) {
                if (order != null) {
                    totalRevenue += order.getTotalAmount();
                    totalOrders++;
                }
            }

            System.out.println("Общее количество заказов: " + totalOrders);
            System.out.printf("Общая выручка: %.2f руб.%n", totalRevenue);
            
            if (totalOrders > 0) {
                System.out.printf("Средний чек: %.2f руб.%n", (totalRevenue / totalOrders));
            }

            System.out.println();
            System.out.println("Детали заказов:");
            
            for (Order order : orders) {
                if (order != null) {
                    System.out.println("  Заказ #" + order.getOrderId() + 
                                     " - Сумма: " + order.getTotalAmount() + " руб." +
                                     " - Блюд: " + order.getMenuItemsCount());
                }
            }
        }

        System.out.println("=========================");
    }

    // Управление складом (вызов отчета)
    public void manageInventory(Inventory inventory) {
        if (inventory != null) {
            System.out.println("Управление складом менеджером: " + getFullName());
            inventory.printInventoryReport();
            System.out.println();
            inventory.printExpiryDates();
        }
    }
    
    // Переопределение метода расчета зарплаты (без вызова базового метода)
    @Override
    public double calculateSalary() {
        // Переопределение без вызова super.calculateSalary()
        // Используем protected-поля из базового класса напрямую
        double base = hoursWorked * hourlyRate;
        double managementBonus = base * 0.2;  // 20% надбавка менеджера
        double total = base + managementBonus + salaryBalance;
        
        System.out.println("[Manager] Расчет зарплаты: базовая = " + base
                + ", надбавка менеджера = " + managementBonus
                + ", баланс = " + salaryBalance);
        
        return total;
    }
    
    // Реализация методов интерфейса IReportGenerator
    @Override
    public String generateReport() {
        StringBuilder sb = new StringBuilder();
        sb.append("Отчет менеджера: ").append(getFullName()).append("\n");
        sb.append("Должность: ").append(getPosition()).append("\n");
        sb.append("Отработано часов: ").append(getHoursWorked()).append("\n");
        sb.append("Почасовая ставка: ").append(getHourlyRate()).append(" руб./час\n");
        sb.append("Зарплата: ").append(calculateSalary()).append(" руб.\n");
        sb.append("Количество продаж: ").append(getSalesCount());
        return sb.toString();
    }
    
    @Override
    public String getReportType() {
        return "Отчет менеджера";
    }
    
    // Реализация методов интерфейса ICloneable
    @Override
    public Manager cloneShallow() {
        // Поверхностное клонирование
        Manager cloned = new Manager();
        // Копируем protected-поля из базового класса
        cloned.fullName = this.fullName;
        cloned.age = this.age;
        cloned.contactNumber = this.contactNumber;
        cloned.address = this.address;
        cloned.position = this.position;
        cloned.hoursWorked = this.hoursWorked;
        cloned.hourlyRate = this.hourlyRate;
        cloned.salaryBalance = this.salaryBalance;
        cloned.salesCount = this.salesCount;
        return cloned;
    }
    
    @Override
    public Manager cloneDeep() {
        // Глубокое клонирование (для Manager совпадает с shallow, так как нет сложных объектов)
        return cloneShallow();
    }
    
    // Метод для демонстрации использования абстрактного класса
    public void printReportUsingAbstractClass() {
        reportGenerator.printReport();
    }
    
    // Метод для присваивания объекта базового класса (аналог перегрузки оператора присваивания)
    public Manager assignFromEmployee(Employee other) {
        if (other != null) {
            // Копируем доступные поля из базового класса
            this.fullName = other.getFullName();
            this.age = other.getAge();
            this.contactNumber = other.getContactNumber();
            this.address = other.getAddress();
            this.position = other.getPosition();
            this.hoursWorked = other.getHoursWorked();
            this.hourlyRate = other.getHourlyRate();
            this.salaryBalance = other.getSalaryBalance();
            this.salesCount = other.getSalesCount();
        }
        return this;
    }
}

